clientd3d/music.c

```diff
@@ -1,853 +1,257 @@
-      }
-   }
+   char *filename;
+   char fname[MAX_PATH + FILENAME_MAX];
 
-   mciSeekParms.dwTo = music_pos;
-   mciSendCommand(midi_bg_music_element, MCI_SEEK,
-                  MCI_TO, (DWORD_PTR)(LPVOID) &mciSeekParms);
-   // Unpause the music!
-   mciPlayParms.dwFrom = music_pos;
-   debug(("Music Position:%ld\n", music_pos));
+   debug(("PlayMusicRsc %d\n", rsc));
+   /* Begin looping sound transition: mark current looping sounds so newly-started
+      sounds for the new room can be registered and protected. */
+   Sound_BeginLoopingSoundTransition();
 
-   if (playing_midi)
+   if (rsc == 0)
    {
-      /* Stop the Jala Song first! */
-      debug(("Unpausing, stopping MIDI\n"));
-      mciSendCommand(midi_element, MCI_STOP, 0, 0);
-      if (dwReturn = mciSendCommand(midi_element, MCI_CLOSE, 0, 0))
-      {
-         debug(("mciClose problem for midi_element\n"));
-         mciGetErrorString(dwReturn, temp, 80);
-         debug((temp));
-         debug((" \n"));
-      }
-      else
-      {
-         playing_midi = false;
-      }
+      // Stop any current background music only; leave ambient sounds alone.
+      MusicStop();
+      playing_music = false;
+      bg_music = 0;
+      latest_music = 0;
+      current_music_file[0] = '\0';
+      return;
    }
 
-   isMusicPaused = false;
-   mciSendCommand(midi_bg_music_element, MCI_PLAY, MCI_FROM | MCI_NOTIFY, (DWORD_PTR)(LPVOID) &mciPlayParms);
-
-   debug(("Unpausing to  position = %ld\n", music_pos));
-#endif
-}
-/******************************************************************************/
-/*
- * PlayMidiRsc:  Play MIDI file associated with given resource number.
- *
- * This function handles pausing or stopping music and midi files and
- * then posts a message which is handled and sends us to NewMusic.
- */
-void PlayMidiRsc(ID rsc)
-{
-   debug(("PlayMidiRsc %d, bg_music is:%d \n", rsc, bg_music));
-   // Save the rsc as latest_music in case our music is off in the config
-   // This way if we toggle it on we have the correct rsc to play.
    latest_music = rsc;
+   bg_music = rsc;
 
-   if (!config.play_music || !has_midi)      
+   if (!config.play_music)
       return;
-   // playing_music is still true when we are paused 
-   // so check both status before pausing
-   if (playing_music && !isMusicPaused)
-   {
-      paused_music = bg_music;
-      PauseMusic();
-#ifndef M59_MSS
-      // Kill the current Midi then POST - NewMusic type = midi.
-      mciSendCommand(midi_element, MCI_CLOSE, 0, 0);
-      playing_midi = false;
-#endif
-      PostMessage(hMain, BK_NEWSOUND, SOUND_MIDI, rsc);
+
+   // Look up the filename for this resource
+   if ((filename = LookupNameRsc(rsc)) == NULL)
       return;
-   }
 
-#ifndef M59_MSS
-   DWORD dwReturn;
-   /* If NOT playing music and IF playing midi...*/
-   /* Stop the midi and prepare to restart a new one. */
-   if (playing_midi)
+   snprintf(fname, sizeof(fname), "%s\\%.*s", music_dir, FILENAME_MAX, filename);
```

Comment:
If you use std::filename::path, there's no messing with backslashes, etc.

https://en.cppreference.com/w/cpp/filesystem/path/append.html